src/amo/pages/CategoryPage/index.js

```diff
@@ -0,0 +1,146 @@
+
+type Props = {|
+  match: {|
+    ...ReactRouterMatchType,
+    params: {| categorySlug: string, visibleAddonType: string |},
+  |},
+|};
+
+type PropsFromState = {|
+  addonType: string,
+  categoryName: string | null,
+  filters: SearchFilters,
+|};
+
+type InternalProps = {|
+  ...Props,
+  ...PropsFromState,
+  dispatch: DispatchFunc,
+  errorHandler: ErrorHandlerType,
+  i18n: I18nType,
+|};
+
+export class CategoryPageBase extends React.Component<InternalProps> {
+  getPageTitle(categoryName: string | null): string {
+    const { addonType, i18n } = this.props;
+
+    if (addonType === ADDON_TYPE_STATIC_THEME) {
+      return categoryName
+        ? i18n.sprintf(i18n.gettext('Themes in %(categoryName)s'), {
+            categoryName,
+          })
+        : i18n.gettext('Themes');
+    }
+    return categoryName
+      ? i18n.sprintf(i18n.gettext('Extensions in %(categoryName)s'), {
+          categoryName,
+        })
+      : i18n.gettext('Extensions');
+  }
+
+  constructor(props: InternalProps) {
+    super(props);
+
+    if (!this.props.categoryName) {
+      this.props.dispatch(
+        fetchCategories({ errorHandlerId: this.props.errorHandler.id }),
+      );
+    }
+  }
+
+  render(): React.Node {
+    const { addonType, categoryName, filters, match } = this.props;
+    const { categorySlug } = match.params;
+
+    const filtersForSearch = {
+      ...filters,
+      addonType,
+      category: categorySlug,
+      sort: filters.sort || `${SEARCH_SORT_RECOMMENDED},${SEARCH_SORT_POPULAR}`,
+    };
+
+    return (
+      <Page>
+        <HeadLinks />
+        <Search
+          enableSearchFilters
+          filters={filtersForSearch}
+          pageTitle={this.getPageTitle(categoryName)}
+          paginationQueryParams={convertFiltersToQueryParams(filters)}
+          pathname={getCategoryResultsPathname({
+            addonType,
+            slug: categorySlug,
+          })}
+        />
+      </Page>
+    );
+  }
+}
+
+const mapStateToProps = (state: AppState, ownProps: Props): PropsFromState => {
+  const { location } = state.router;
+  const { categorySlug, visibleAddonType } = ownProps.match.params;
+
+  let categoryName = null;
+
+  const addonType = apiAddonType(visibleAddonType);
+  const categoriesState = state.categories.categories;
+  const { clientApp } = state.api;
+
+  if (categoriesState && clientApp) {
+    const appTypes = categoriesState[clientApp];
+
+    if (appTypes) {
+      const categories = appTypes[addonType];
+      categoryName = getCategoryName(categories, categorySlug);
+    }
+  }
+
+  const filtersFromLocation = convertQueryParamsToFilters(location.query);
+
+  return {
+    addonType,
+    categoryName,
+    filters: fixFiltersFromLocation(filtersFromLocation),
+  };
+};
+
+const CategoryPage: React.ComponentType<Props> = compose(
+  connect(mapStateToProps),
+  translate(),
+  withErrorHandler({ name: 'CategoryPage' }),
+)(CategoryPageBase);
+
+export default CategoryPage;
```

Comment:
Shouldn't we return a 404 when the category slug is invalid? For example: http://localhost:3000/en-US/firefox/themes/category/nope/

Reply:
Good catch. I assumed the logic in `Page` would take care of that, but the API error is coming from the `Search` component. I think the best way to address this is to just change the logic of the `Search` component to display a 404 page if the API response is either `404` or `400`. I can't think of an occasion on which we'd want to display anything other than a 404 page when the search API returns a 400. I've made this change and you can see what you think when I push the commit.

Reply:
I think it's fine (and it will be even better with a comment) but we should probably file an issue because for other search parameters, we display the error on the search page, which is kinda nice too.